# Single-cell
## Load data
### Barcodes.tsv.gz, features.tsv.gz, matrix.mtx.gz
1.Barcodes.tsv.gz: cell label(colnames) 
2.Features.tsv.gz: gene id(rownames=umi) 
3.Matrix.mtx.gz: expression data.

These three files are generated by cellranger, which is used to align reads and generate feature-barcode matrices
```{r}
library(dplyr)
library(Seurat)
library(patchwork)
# Load the PBMC dataset, seurat uses a sparse-matrix to save memory and enhance speed.（object.size）
count <- Read10X(data.dir = "./exampledata/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <-
  CreateSeuratObject(
    counts = count,
    project = "pbmc3k",
    min.cells = 0,
    min.features = 0
  )
```
### Count file
```{r}
scRNA <- CreateSeuratObject(counts = counts)
```
### H5 file
H5 is a data file saved in the Hierarchical Data Format (HDF). It contains multidimensional arrays of data.
```{r}
sce <- Read10X_h5(filename = xxx_matrices_h5.h5")
sce <- CreateSeuratObject(counts = sce)
```
### H5ad file
1.convert h5ad to h5seurat
2.load seurat object by "loadh5seurat" function
```{r}
library(SeuratDisk)
Convert("xxx_raw_counts.h5ad", "h5seurat",
        overwrite = TRUE,assay = "RNA")
scRNA <- LoadH5Seurat("GSE153643_RAW/GSM4648565_liver_raw_counts.h5seurat")
```

## Standard pre-processing workflow
GOAL:  To ensure that only single, live cells are included in downstream analysis(doi: 10.1186/s13059-016-0888-1)
### QC
#### The number of unique genes and total molecules

#### Filter out low-quality/dying cells through the percentage of mitochondrial genome in all genome

Low-quality / dying cells often exhibit extensive mitochondrial contamination

Using `PercentageFeatureSet()` function to calculate the percentage of mitochondrial gene in cell, which be able to identify dying cells or low- quality cells.

(Sum of the counts for features belonging to the set)*100 / the column sum for all features times (https://www.rdocumentation.org/packages/Seurat/versions/3.1.3/topics/PercentageFeatureSet)
```{r mito}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```
